cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java

```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
   @POST
   @Path("/metadata")
   public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
-    String requestBody = request.content().toString(StandardCharsets.UTF_8);
+    Map<String, GenericRecord> consumeRequests = decodeConsumeRequest(request);
+    MessagesBodyProducer messagesBodyProducer = new MessagesBodyProducer(cConf, consumeRequests, messageFetcher);
+    responder.sendContent(HttpResponseStatus.OK, messagesBodyProducer,
+                          new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE, "avro/binary"));
+  }
 
-    Map<String, MonitorConsumeRequest> consumeRequests = GSON.fromJson(requestBody, MAP_STRING_CONSUME_REQUEST_TYPE);
+  /**
+   * shuts down remote runtime
+   */
+  @POST
+  @Path("/shutdown")
+  public void shutdown(FullHttpRequest request, HttpResponder responder) throws Exception {
+    responder.sendString(HttpResponseStatus.OK, "Triggering shutdown down Runtime Http Server.");
+    shutdownRunnable.run();
+  }
 
-    ChunkResponder chunkResponder = responder.sendChunkStart(
-      HttpResponseStatus.OK, new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE,
-                                                          "application/json; charset=utf-8"));
-    ByteBuf buffer = Unpooled.buffer();
+  /**
+   * Decode consume request from avro binary format
+   */
+  private Map<String, GenericRecord> decodeConsumeRequest(FullHttpRequest request) throws IOException {
+    Decoder decoder = DecoderFactory.get().directBinaryDecoder(new ByteBufInputStream(request.content()), null);
+    DatumReader<Map<String, GenericRecord>> datumReader = new GenericDatumReader<>(
+      MonitorSchemas.V1.MonitorConsumeRequest.SCHEMA);
+    return datumReader.read(null, decoder);
+  }
 
-    try (JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(new ByteBufOutputStream(buffer),
-                                                                       StandardCharsets.UTF_8))) {
-      jsonWriter.beginObject();
 
-      for (Map.Entry<String, MonitorConsumeRequest> entry : consumeRequests.entrySet()) {
-        String topicConfig = entry.getKey();
-        jsonWriter.name(topicConfig);
-        jsonWriter.beginArray();
+  /**
+   * A {@link BodyProducer} to encode and send back messages.
+   * Instead of using GenericDatumWriter, we perform the map and array encoding manually so that we don't have to buffer
+   * all messages in memory before sending out.
+   */
+  private static class MessagesBodyProducer extends BodyProducer {
+    private final Iterator<Map.Entry<String, GenericRecord>> requestsIterator;
+    // caches request key to topic
+    private final Map<String, String> requestKeyToLocalTopic;
```

Comment:
This map seems unnecessary since for each key from the request, you only have to resolve the local topic once and never get reused, right?